### PR TITLE
BS-4: Implement DLR handling

### PR DESCRIPTION
**Pull request by abdulazizali77**
**What this PR does / why we need it**:
- Implement DLR handling
- Outbound SMSs will have status updated after DLR returns

- Add SMSMessageId to SMPPOutboundMessageEntity
- Make submit_sm async and listen for incoming submit_sm_resp
- Save memory hashMap of resp_msg_id to SMSMessagId
- Handle incoming DLR and update DAO layer

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # BS-30

**Special notes for your reviewer**:

### CODE REVIEW

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/SmsSession.java [hunk 1]
```diff
@@ -235,6 +236,13 @@ public void onReceive(final Object message) throws Exception {
             final SmsSessionAttribute attribute = (SmsSessionAttribute) message;
             attributes.put(attribute.name(), attribute.value());
             Object record = attributes.get("record");
+            try {
```
Comment (jaimecasero) [inline restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/SmsSession.java]: is this try necesasary to test the casting, im not sure i like the construct i would prefer to use instanceof

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/SmsSession.java [hunk 2]
```diff
@@ -350,8 +360,17 @@ private boolean sendUsingSmpp(String from, String to, String body, TlvSet tlvSet
                 logger.info("SMPP session is available and connected, outbound message will be forwarded to :  " + to );
                 logger.info("Encoding:  " + encoding );
             }
+
+            SmsMessage record = (SmsMessage)this.attributes.get("record");
+            Sid sid = null;
+            if(record!=null) {
+                sid = record.getSid();
+                logger.error("record sid = "+sid.toString());
```
Comment (jaimecasero) [inline restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/SmsSession.java]: lets be cautius with logger level. I dont thin kthis is error level

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppClientOpsThread.java [hunk 3]
```diff
@@ -411,7 +447,34 @@ public PduResponse firePduRequestReceived(PduRequest pduRequest) {
...
+                       }
+
+                       //TODO: point to an array of provider specific functions
+                       Map<String,String> dlr = getMessageIdNexmo(decodedPduMessage);
+                       String messageId = dlr.get("id");
+                       String status = dlr.get("status");
+                       DateTime dateSent = null;
+                       try {
+                           //TODO: abstract Nexmo specific format
+                           dateSent = DateTime.parse(dlr.get("sent_date"), DateTimeFormat.forPattern("yyMMddkkmm"));
```
Comment (jaimecasero) [inline restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppClientOpsThread.java]: can we save the pattern to be reused?, I guess compiling the pattern is resource consuming...

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppMessageHandler.java [hunk 4]
```diff
@@ -129,6 +135,38 @@ public void onReceive(Object message) throws Exception {
...
+            SmsMessagesDao smsMessagesDao = storage.getSmsMessagesDao();
+            SmsMessage smsMessage = smsMessagesDao.getSmsMessage(smppDeliveryReceiptEntity.getDaoMessageSid());
+            //QUEUED("queued"), SENDING("sending"), SENT("sent"), FAILED("failed"), RECEIVED("received");
+            //"id:XXXXXXXXXX sub:001 dlvrd:000 submit date:YYMMDDHHMM done date:YYMMDDHHMM stat:ZZZZZZZ err:YYY text:none
+            //FIXME: this is Nexmo specific, we need to create spi and add various Providers
+            SmsMessage.Status status = null;
+            DateTime dateSent =smppDeliveryReceiptEntity.getDateSent();
+            //TODO: modify enum so we can do away with the switch
+            switch(smppDeliveryReceiptEntity.getStatus().toLowerCase()) {
+            case "queued":
```
Comment (jaimecasero) [inline restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppMessageHandler.java]: can we replace this with a Map doing the mapping

**Comment by abdulazizali77** [comment]
@jaimecasero have addressed comments, PTAL

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppMessageHandler.java [hunk 5]
```diff
@@ -129,6 +132,20 @@ public void onReceive(Object message) throws Exception {
...
+            SmsMessage smsMessage = smsMessagesDao.getSmsMessage(smppDeliveryReceiptEntity.getDaoMessageSid());
+            SmsMessage.Status status = smppDeliveryReceiptEntity.getStatus();
+            DateTime dateSent =smppDeliveryReceiptEntity.getDateSent();
+
+            if(logger.isInfoEnabled()) {
+                logger.info("updating "+smsMessage.getSid()+" status="+status+" dateSent="+dateSent);
+            }
+
+            if(smsMessage != null && status != null) {
+                smsMessagesDao.updateSmsMessage(smsMessage.setStatus(status).setDateSent(dateSent));
```
Comment (jaimecasero) [inline restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppMessageHandler.java]: is this dateSent, or dateDelivered? not sure if we have a particular Date field for this...

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppDeliveryReceiptEntity.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
+package org.restcomm.connect.sms.smpp;
+
+import org.joda.time.DateTime;
+import org.restcomm.connect.commons.dao.Sid;
+import org.restcomm.connect.dao.entities.SmsMessage;
+
+public class SmppDeliveryReceiptEntity {
```
Comment (jaimecasero) [inline restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppDeliveryReceiptEntity.java]: is this an entity, or simply a Mesasge to pass DTO around Akka actors...?

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/dlr/provider/TelestaxDlrParser.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        statusMap.put("DELIVRD", SmsMessage.Status.DELIVERED);
+        statusMap.put("UNKNOWN", SmsMessage.Status.SENDING);
+    }
+
+    @Override
+    public Map<String, String> parseMessage(String message) {
+        //"id:XXXXXXXXXX sub:001 dlvrd:000 submit date:YYMMDDHHMM done date:YYMMDDHHMM stat:ZZZZZZZ err:YYY text:none
+        Map<String,String> dlr = new HashMap<String, String>();
+
+        final String idTag = "id:";
```
Comment (jaimecasero) [inline restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/dlr/provider/TelestaxDlrParser.java]: can we define these as CONSTANTS? do we have these contants defined in stack in another placem so we can reuse?

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/dlr/provider/TelestaxDlrParser.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        final String statTag = "stat:";
+        final int idStartPos = message.indexOf(idTag)+idTag.length();
+        final int submitStartPos = message.indexOf(submitTag)+submitTag.length();
+        final int doneStartPos = message.indexOf(doneTag)+doneTag.length();
+        final int statStartPos = message.indexOf(statTag)+statTag.length();
+        final int idEndPos = message.indexOf(" ",idStartPos);
+        final int submitEndPos = message.indexOf(" ",submitStartPos);
+        final int doneEndPos = message.indexOf(" ",doneStartPos);
+        final int statEndPos = message.indexOf(" ",statStartPos);
+        final String messageId  = message.substring(idStartPos, idEndPos);
```
Comment (jaimecasero) [inline restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/dlr/provider/TelestaxDlrParser.java]: should we check positions detected so we dont fall into stringoutofbounds situations..?

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/dlr/spi/DlrParser.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package org.restcomm.connect.sms.smpp.dlr.spi;
+
+import java.util.Map;
+
+import org.joda.time.DateTime;
+import org.restcomm.connect.dao.entities.SmsMessage;
+
+public interface DlrParser {
+    Map<String, String> parseMessage(String message);
+    DateTime getDate(String message);
```
Comment (jaimecasero) [inline restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/dlr/spi/DlrParser.java]: i would use other verbs here than "get", if they are really "parsing"...

File: restcomm/restcomm.sms/src/test/java/org/restcomm/connect/sms/smpp/dlr/provider/TelestaxDlrParserTest.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+		DlrParser parser = new TelestaxDlrParser();
+		Map<String,String> dlrMap = parser.parseMessage(pduMessage);
+		Assert.assertEquals("0000058049", dlrMap.get("id"));
+		Assert.assertEquals("DELIVRD", dlrMap.get("status"));
+		Assert.assertEquals("1805170144", dlrMap.get("sent_date"));
+		Assert.assertEquals("1805170144", dlrMap.get("submit_date"));
+	}
+	
+	@Test
+	public void getRestcommStatus() throws SmppInvalidArgumentException {
```
Comment (jaimecasero) [inline restcomm/restcomm.sms/src/test/java/org/restcomm/connect/sms/smpp/dlr/provider/TelestaxDlrParserTest.java]: i dont get this test. where is statusMap defined, and used? wy do we always cehck same states(DELIVERED)?

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppClientOpsThread.java [hunk 11]
```diff
@@ -352,19 +391,11 @@ public void firePduRequestExpired(PduRequest pduRequest) {
...
             PduResponse response = pduRequest.createResponse();
-            //Nexmo keeps sending enquire_link in response that causes
-            //problem with normal PDU request. This tells Restcomm SMPP to do Nothing
-            //with the request
-            if (pduRequest.toString().toLowerCase().contains("enquire_link")) {
-                //logger.info("This is a response to the enquire_link, therefore, do NOTHING ");
-            } else {
 
-                DeliverSm deliverSm = (DeliverSm) pduRequest;
+            if (!pduRequest.toString().toLowerCase().contains("enquire_link")) {
```
Comment (jaimecasero) [inline restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppClientOpsThread.java]: i would like this condition to be more accurate to do the DeliverSm casting later..

**Pull request merged**